shared/src/ledger/eth_bridge/storage/bridge_pool.rs

```diff
@@ -44,3 +58,847 @@ pub fn is_bridge_pool_key(key: &Key) -> bool {
 pub fn is_protected_storage(key: &Key) -> bool {
     is_bridge_pool_key(key) && *key != get_pending_key()
 }
+
+/// A simple Merkle tree for the Ethereum bridge pool
+#[derive(
+    Debug, Default, Clone, BorshSerialize, BorshDeserialize, BorshSchema,
+)]
+pub struct BridgePoolTree {
+    /// Root of the tree
+    root: KeccakHash,
+    /// The underlying storage
+    store: BTreeSet<KeccakHash>,
+}
+
+impl BridgePoolTree {
+    /// Create a new merkle tree for the Ethereum bridge pool
+    pub fn new(root: KeccakHash, store: BTreeSet<KeccakHash>) -> Self {
+        Self { root, store }
+    }
+
+    /// Parse the key to ensure it is of the correct type.
+    ///
+    /// If it is, it can be converted to a hash.
+    /// Checks if the hash is in the tree.
+    pub fn contains_key(&self, key: &Key) -> Result<bool, Error> {
+        Ok(self.store.contains(&Self::parse_key(key)?))
+    }
+
+    /// Update the tree with a new value.
+    ///
+    /// Returns the new root if successful. Will
+    /// return an error if the key is malformed.
+    pub fn update_key(&mut self, key: &Key) -> Result<Hash, Error> {
+        let hash = Self::parse_key(key)?;
+        _ = self.store.insert(hash);
+        self.root = self.compute_root();
+        Ok(self.root())
+    }
+
+    /// Delete a key from storage and update the root
+    pub fn delete_key(&mut self, key: &Key) -> Result<(), Error> {
+        let hash = Self::parse_key(key)?;
+        _ = self.store.remove(&hash);
+        self.root = self.compute_root();
+        Ok(())
+    }
+
+    /// Compute the root of the merkle tree
+    pub fn compute_root(&self) -> KeccakHash {
+        let mut hashes: Vec<KeccakHash> = self.store.iter().cloned().collect();
+        while hashes.len() > 1 {
+            let mut next_hashes = vec![];
+            let left_leaves = hashes.iter().step_by(2);
+            let mut right_leaves = hashes.iter();
+            _ = right_leaves.next();
+            let mut right_leaves = right_leaves.step_by(2);
+
+            for left in left_leaves {
+                let right = right_leaves.next().cloned().unwrap_or_default();
+                next_hashes.push(hash_pair(left.clone(), right));
+            }
+            hashes = next_hashes;
+        }
+
+        if hashes.is_empty() {
+            Default::default()
```

Comment:
We have to do this. The root of an empty tree is by definition `[0u8; 32]`. We can't have an option, otherwise we can't compute the root of our total merkle store in a way that is committing to this subtree and we lose security properties.